### PR TITLE
Implement hash validation for Tezos

### PR DESCRIPTION
**Pull request by CapnMigraine**
Add validation for address, block and transaction hashes
Validate encoding, prefix and length

In Tezos, hashes are encoded using Base58Check

Ticket: BG-17820

### CODE REVIEW

File: src/coin/xtz/iface.ts [hunk 1]
```diff
@@ -32,3 +32,8 @@ export function isPrivateKey(source: KeyPairOptions): source is PrivateKey {
 export function isPublicKey(source: KeyPairOptions): source is PublicKey {
   return (source as PublicKey).pub !== undefined;
 }
+
+export interface HashType {
+  prefix: Buffer;
+  bytes: number;
```
Comment (argjv) [inline src/coin/xtz/iface.ts]: s/bytes/byteLength/g

File: src/coin/xtz/iface.ts [hunk 2]
```diff
@@ -32,3 +32,8 @@ export function isPrivateKey(source: KeyPairOptions): source is PrivateKey {
 export function isPublicKey(source: KeyPairOptions): source is PublicKey {
   return (source as PublicKey).pub !== undefined;
 }
+
+export interface HashType {
```
Comment (argjv) [inline src/coin/xtz/iface.ts]: JSDoc or add comments next to the fields. Does `byte` include the prefix? Is this the decoded bytes length?

File: src/coin/xtz/utils.ts [hunk 3]
```diff
@@ -1,82 +1,250 @@
...
-  const textPrefix = decodedText.slice(0, prefix.length);
-  // Check prefix
-  if (!textPrefix.equals(Buffer.from(prefix))) {
+/**
+ * Decode a string encoded as base58
+ *
+ * @param {String} hash - the string to decode
+ * @returns {Buffer | undefined}
+ */
+export function base58decode(hash): Buffer | undefined {
```
Comment (argjv) [inline src/coin/xtz/utils.ts]: Missing input type

File: src/coin/xtz/utils.ts [hunk 4]
```diff
@@ -1,82 +1,250 @@
...
+/**
+ * Returns whether or not the string is a valid Tezos hash of the given type
+ *
+ * @param {String} hash - the string to validate
+ * @param {HashType} hashType - the type of the provided hash
+ * @returns {Boolean}
+ */
+export function isValidHash(hash: string, hashType: HashType): boolean {
+  // Validate encoding
+  const decodedHash = base58decode(hash);
```
Comment (argjv) [inline src/coin/xtz/utils.ts]: I'm not sure we need `base58decode`, I don't see it being referenced anywhere else. We could just do:

```
  let decodedHash;
  try {
    decodedHash = base58check.decode(hash);
  } catch (e) {
    return false;
  }
```

File: src/coin/xtz/iface.ts (hunk 2 above)
Comment (CapnMigraine) [reply src/coin/xtz/iface.ts]: Added comments for clarity

File: src/coin/xtz/iface.ts (hunk 1 above)
Comment (CapnMigraine) [reply src/coin/xtz/iface.ts]: done

File: src/coin/xtz/utils.ts (hunk 3 above)
Comment (CapnMigraine) [reply src/coin/xtz/utils.ts]: I removed the function, as per your suggestion in the comment below

File: src/coin/xtz/utils.ts (hunk 4 above)
Comment (CapnMigraine) [reply src/coin/xtz/utils.ts]: You're right, I added it as a more generic helper, but these utility functions are Tezos-specific, so I removed the function

File: src/coin/xtz/utils.ts [hunk 5]
```diff
@@ -41,7 +25,12 @@ export function base58decode(hash): Buffer | undefined {
...
 export function isValidHash(hash: string, hashType: HashType): boolean {
   // Validate encoding
-  const decodedHash = base58decode(hash);
+  let decodedHash;
+  try {
+    decodedHash = base58check.decode(hash);
+  } catch (e) {
+    return false;
+  }
   if (!decodedHash) {
```
Comment (argjv) [inline src/coin/xtz/utils.ts]: Can `base58check.decode(hash);` return undefined? If not, this `if` block is not neccesary.

Comment (CapnMigraine) [reply src/coin/xtz/utils.ts]: it's not needed, removed

**Pull request merged**